rust/lance/src/index/vector/ivf/shuffler.rs

```diff
@@ -0,0 +1,194 @@
+        for (key, batches) in self.buffer.iter() {
+            if !batches.is_empty() {
+                self.parted_groups
+                    .entry(*key)
+                    .or_default()
+                    .push(self.writer.batch_id as u32);
+                self.writer.write(batches.as_slice()).await?;
+            }
+        }
+        self.writer.finish().await?;
+        Ok(Shuffler::new(&self.parted_groups, &self.temp_dir))
+    }
+}
+
+pub struct Shuffler<'a> {
+    /// Partition ID to file-group ID mapping, in memory.
+    /// No external dependency is required, because we don't need to guarantee the
+    /// persistence of this mapping, as well as the temp files.
+    parted_groups: BTreeMap<u32, Vec<u32>>,
+
+    /// We need to keep the temp_dir with Shuffler because ObjectStore crate does not
+    /// work with a NamedTempFile.
+    temp_dir: &'a TempDir,
+}
+
+impl<'a> Shuffler<'a> {
+    #[allow(dead_code)]
+    fn new(parted_groups: &BTreeMap<u32, Vec<u32>>, temp_dir: &'a TempDir) -> Self {
+        Self {
+            parted_groups: parted_groups.clone(),
+            temp_dir,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn keys(&self) -> Vec<u32> {
+        self.parted_groups.keys().copied().collect()
+    }
+
+    /// Iterate over the shuffled [RecordBatch]s for a given partition key.
+    #[allow(dead_code)]
+    pub async fn key_iter(&self, key: u32) -> Result<impl RecordBatchStream + '_> {
+        let file_path = lance_buffer_path(self.temp_dir)?;
+        let (object_store, _) = ObjectStore::from_uri("file://").await?;
+        let reader = FileReader::try_new(&object_store, &file_path).await?;
+        let schema = reader.schema().clone();
+
+        let group_ids = self
+            .parted_groups
+            .get(&key)
+            .unwrap_or(&vec![])
+            .iter()
+            .copied()
+            .collect::<HashSet<_>>();
+        let stream = batches_stream(reader, schema, move |id| group_ids.contains(&(*id as u32)));
+        Ok(stream)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::sync::Arc;
+
+    use arrow_array::UInt32Array;
+    use arrow_schema::{DataType, Field, Schema};
+    use futures::TryStreamExt;
+
+    #[tokio::test]
+    async fn test_shuffler() {
+        let schema = Schema::new(vec![Field::new("a", DataType::UInt32, false)]);
+        let mut shuffler = ShufflerBuilder::try_new(&schema, 4).await.unwrap();
+        for i in 0..20 {
+            shuffler
+                .insert(
+                    i % 3,
+                    RecordBatch::try_new(
+                        Arc::new(schema.clone()),
+                        vec![Arc::new(UInt32Array::from(vec![i]))],
+                    )
+                    .unwrap(),
+                )
+                .await
+                .unwrap();
+        }
+        let reader = shuffler.finish().await.unwrap();
+        assert_eq!(reader.keys(), vec![0, 1, 2]);
+        for i in 0..3 {
```

Comment:
Could we add a test for what happens when you request group id = 3 (where there isn't any entry?)

Reply:
Done